openapi/spec-generator/specbuilder.go

```diff
@@ -0,0 +1,490 @@
+	if err != nil {
+		return err
+	}
+
+	b.resolveRefPaths()
+
+	return nil
+}
+
+func (b *SpecBuilder) buildGraphFromModel(api v1api) error {
+	paths := openapi3.Paths{}
+
+	for _, path := range api.GetPaths() {
+		if len(path.Template) < 1 {
+			err := fmt.Errorf("invalid Path spec: %#v", path)
+			b.logger(err)
+			return err
+		}
+
+		pathItem := &openapi3.PathItem{}
+
+		for _, op := range path.Operations {
+			var err error
+			var requestBodyRef *openapi3.RequestBodyRef
+			var responses openapi3.Responses
+
+			if op.RequestBody != nil {
+				requestBodyRef, err = b.adaptRequestBody(op.RequestBody)
+				if err != nil {
+					err = fmt.Errorf("error %s: unable to adapt request body for %#v", err, op.RequestBody)
+					b.logger(err)
+					return err
+				}
+				if requestBodyRef != nil {
+					requestBodyRef = &openapi3.RequestBodyRef{
+						Ref:   "",
+						Value: requestBodyRef.Value,
+					}
+				}
+			}
+			if responses, err = b.adaptResponses(op.Responses); err != nil {
+				err = fmt.Errorf("error %s: unable to adapt Operation responses for %#v", err, op)
+				b.logger("unable to adapt Operation responses", err, path, op)
+				return err
+			}
+
+			operation := &openapi3.Operation{
+				Tags:        op.Tags,
+				Summary:     op.Summary,
+				Description: op.Description,
+				OperationID: op.Description,
+				Parameters:  b.adaptParameters(op.Parameters),
+				RequestBody: requestBodyRef,
+				Responses:   responses,
+			}
+
+			// Add security if required
+			security := b.adaptSecurityRequirements(op.Parameters)
+			if security != nil {
+				operation.Security = security
+			}
+
+			switch op.Method {
+			case http.MethodGet:
+				pathItem.Get = operation
+			case http.MethodPost:
+				pathItem.Post = operation
+			case http.MethodDelete:
+				pathItem.Delete = operation
+			}
+		}
+		paths[path.Template] = pathItem
+	}
+
+	b.spec.Model.Paths = paths
+
+	return nil
+}
+
+func (b *SpecBuilder) adaptParameters(params []*Parameter) openapi3.Parameters {
+	parameters := openapi3.Parameters{}
+
+	for _, param := range params {
+		if existing, ok := b.spec.Model.Components.Parameters[param.Id]; ok {
+			parameters = append(parameters, existing)
+		} else {
+			parameter := b.addParameter(param)
+			parameters = append(parameters, parameter)
+		}
+	}
+
+	return parameters
+}
+
+func (b *SpecBuilder) addParameter(param *Parameter) *openapi3.ParameterRef {
+	if existing, ok := b.spec.Model.Components.Parameters[param.Id]; ok {
+		return existing
+	}
+
+	parameter := &openapi3.ParameterRef{
+		Ref: "",
+		Value: &openapi3.Parameter{
+			Name:        param.Name,
+			Description: param.Description,
+			Schema: openapi3.NewSchemaRef("", &openapi3.Schema{
+				Type: param.SchemaType,
+			}),
+			In:       param.In,
+			Required: param.Required,
+		},
+	}
+
+	b.spec.Model.Components.Parameters[param.Id] = parameter
+
+	return &openapi3.ParameterRef{
+		Ref:   fmt.Sprintf("#/components/parameters/%s", param.Id),
+		Value: parameter.Value,
+	}
```

Comment:
What's the significance for `Ref` and why not return `parameter`? It looks like the first invocation of `addParameter` will return a ParameterRef with Ref set, but if it's invoked again with the same id, the existing cached `parameter` is returned without Ref set.